File: tests/.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cmath>
#include <exception>

template <typename T>
T add(T a, T b) {
    return a + b;
}

class Person {
protected:
    std::string name;
    int age;

public:
    Person(const std::string& name, int age) : name(name), age(age) {
        std::cout << "Person created: " << name << ", age " << age << "\n";
    }

    virtual ~Person() {}

    virtual std::string introduce() const {
        return "Hi, I'm " + name + " and I'm " + std::to_string(age) + " years old.";
    }

    void addHobby(const std::string& hobby) {
        hobbies.push_back(hobby);
    }

    void listHobbies() const {
        for (const auto& hobby : hobbies) {
            std::cout << "Hobby: " << hobby << "\n";
        }
    }

private:
    std::vector<std::string> hobbies;
};

class Employee : public Person {
    std::string position;

public:
    Employee(const std::string& name, int age, const std::string& position)
        : Person(name, age), position(position) {}

    std::string introduce() const override {
        return Person::introduce() + " I work as a " + position + ".";
    }
};

class Shape {
public:
    virtual double area() const = 0;
    virtual double perimeter() const = 0;
    virtual ~Shape() {}
};

class Circle : public Shape {
    double radius;

public:
    Circle(double radius) : radius(radius) {}

    double area() const override {
        return M_PI * radius * radius;
    }

    double perimeter() const override {
        return 2 * M_PI * radius;
    }
};

double divide(double a, double b) {
    if (b == 0) {
        throw std::invalid_argument("Division by zero is not allowed");
    }
    return a / b;
}

int main() {
    try {
        Person john("John", 30);
        std::cout << john.introduce() << "\n";
        john.addHobby("reading");
        john.addHobby("coding");
        john.listHobbies();

        Employee jane("Jane", 28, "Developer");
        std::cout << jane.introduce() << "\n";

        std::cout << "Sum of 3 and 4 (int): " << add(3, 4) << "\n";
        std::cout << "Sum of 3.5 and 4.5 (double): " << add(3.5, 4.5) << "\n";

        Circle circle(5.0);
        std::cout << "Circle: Area = " << circle.area()
                  << ", Perimeter = " << circle.perimeter() << "\n";

        std::cout << "10 / 2 = " << divide(10, 2) << "\n";
        std::cout << "10 / 0 = " << divide(10, 0) << "\n";
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
    }

    std::vector<std::string> names = {"Alice", "Bob", "Charlie"};
    names.push_back("Diana");
    for (const auto& name : names) {
        std::cout << "Name: " << name << "\n";
    }

    return 0;
}

File: tests/.rs
use std::f64::consts::PI;
use std::fmt;
use std::error::Error;

struct Person {
    name: String,
    age: u32,
    hobbies: Vec<String>,
}

impl Person {
    fn new(name: &str, age: u32) -> Self {
        Self {
            name: name.to_string(),
            age,
            hobbies: Vec::new(),
        }
    }

    fn introduce(&self) -> String {
        format!("Hi, I'm {} and I'm {} years old.", self.name, self.age)
    }

    fn add_hobby(&mut self, hobby: &str) {
        self.hobbies.push(hobby.to_string());
    }

    fn list_hobbies(&self) {
        for hobby in &self.hobbies {
            println!("Hobby: {}", hobby);
        }
    }
}

enum Shape {
    Circle(f64),
    Rectangle(f64, f64),
}

impl Shape {
    fn area(&self) -> f64 {
        match self {
            Shape::Circle(radius) => PI * radius * radius,
            Shape::Rectangle(width, height) => width * height,
        }
    }

    fn perimeter(&self) -> f64 {
        match self {
            Shape::Circle(radius) => 2.0 * PI * radius,
            Shape::Rectangle(width, height) => 2.0 * (width + height),
        }
    }
}

fn add<T: std::ops::Add<Output = T>>(a: T, b: T) -> T {
    a + b
}

trait Greet {
    fn greet(&self) -> String;
}

impl Greet for Person {
    fn greet(&self) -> String {
        format!("Hello, my name is {}.", self.name)
    }
}

fn divide(a: f64, b: f64) -> Result<f64, Box<dyn Error>> {
    if b == 0.0 {
        Err("Division by zero is not allowed".into())
    } else {
        Ok(a / b)
    }
}

fn main() {
    let mut john = Person::new("John", 30);
    println!("{}", john.introduce());
    john.add_hobby("reading");
    john.add_hobby("coding");
    john.list_hobbies();
    println!("{}", john.greet());

    let circle = Shape::Circle(5.0);
    let rectangle = Shape::Rectangle(4.0, 7.0);
    println!(
        "Circle: Area = {:.2}, Perimeter = {:.2}",
        circle.area(),
        circle.perimeter()
    );
    println!(
        "Rectangle: Area = {:.2}, Perimeter = {:.2}",
        rectangle.area(),
        rectangle.perimeter()
    );

    println!("Sum of integers: {}", add(3, 4));
    println!("Sum of floats: {:.2}", add(3.5, 4.5));

    match divide(10.0, 2.0) {
        Ok(result) => println!("10 / 2 = {:.2}", result),
        Err(e) => println!("Error: {}", e),
    }

    match divide(10.0, 0.0) {
        Ok(result) => println!("10 / 0 = {:.2}", result),
        Err(e) => println!("Error: {}", e),
    }

    let names = vec!["Alice", "Bob", "Charlie"];
    for (index, name) in names.iter().enumerate() {
        println!("Name #{}: {}", index + 1, name);
    }
}
